src/main.cpp

```diff
@@ -3068,10 +3048,38 @@ bool ActivateBestChain(CValidationState &state, const CChainParams& chainparams,
         }
         // When we reach this point, we switched to a new tip (stored in pindexNewTip).
 
+        // Remove orphan transactions with cs_main
+        {
+            LOCK(cs_main);
```

Comment:
I hope we can separate nodestate and other message-processing related variables from cs_main...